### PR TITLE
[WIP] [tabular] Add Joint Ensembling + Decision Threshold Calibration

### LINKED ISSUE #4119: [tabular] Set `calibrate_decision_threshold="auto"` to get better results on prediction metrics
TabularPredictor can get far better results on metrics such as `f1` and `balanced_accuracy` by specifying `calibrate_decision_threshold=True`. However, currently it is set to `False` by default and most users might overlook setting it to True.

This leads to confusing comparisons such as in the [HyperFast paper](https://arxiv.org/pdf/2402.14335.pdf), where they measure `Balanced Accuracy` showing AutoGluon having not nearly as strong results as when measuring with `Accuracy`. This is because `Accuracy` doesn't normally need decision threshold calibration, as the default (0.5) is near optimal. However, `Balanced Accuracy` often has an optimal threshold far different than 0.5.

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

Resolves #4119

*Description of changes:*

- Implements joint weighted ensemble + decision threshold calibration for better results on compatible metrics.
- Implements fast threshold optimization functions for f1, accuracy, and balanced accuracy metrics. These fast methods are 100s or 1000s of times faster than the black-box grid-search implementation. This is important because the calibration procedure is called N x M times during joint calibration w/ ensembling, where N is the number of iterations and M is the number of model candidates. Given that N is 25 by default, and M can often be 20 or more, this logic can be called 500 times in a single weighted ensemble fit, making it important to be fast.
- Added extensive unit tests for threshold calibration methods to ensure correctness.
- Various code cleanup and doc improvement

TODO:

- [ ] Ensure `sample_weights` are handled correctly in optimized functions.
- [ ] Verify if threshold is re-used for refit_full.
- [ ] Add toggle to enable joint threshold optimization in the TabularPredictor fit call.
- [ ] Consider using optimized function as part of final threshold selection at end of fit.
- [ ] Consider updating all model's val scores to reflect the score achieved with optimal threshold (note: introduces time overhead)
- [ ] Consider implementing additional optimized threshold functions.
- [ ] Consider enabling joint threshold optimization by default.
- [ ] Add a helper function to validate a user-defined custom threshold optimization function based on the logic in the unit tests.
- [ ] Consider adding a plotting utility to visualize the impact of decision threshold on metric score, such as what is done in https://scikit-learn.org/stable/auto_examples/model_selection/plot_cost_sensitive_learning.html#sphx-glr-auto-examples-model-selection-plot-cost-sensitive-learning-py
- [ ] Sklearn [recently implemented threshold calibration](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TunedThresholdClassifierCV.html#sklearn.model_selection.TunedThresholdClassifierCV) in sklearn 1.5. They use stratified cross-validation to tune the threshold. Identify if this is something we want to consider. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |
| botocore==1.34.115 | botocore==1.34.116 |
| fastcore==1.5.41 | fastcore==1.5.42 |
| torchvision==0.18.0 | torchvision==0.17.2 |
| boto3==1.34.115 | boto3==1.34.116 |
| triton==2.3.0 | triton==2.2.0 |
| torch==2.3.0 | torch==2.2.2 |
| nvidia-nccl-cu12==2.20.5 | nvidia-nccl-cu12==2.19.3 |
| botocore==1.34.115 | botocore==1.34.116 |
| fastcore==1.5.41 | fastcore==1.5.42 |
| torchvision==0.18.0 | torchvision==0.17.2 |
| boto3==1.34.115 | boto3==1.34.116 |
| triton==2.3.0 | triton==2.2.0 |
| torch==2.3.0 | torch==2.2.2 |
| nvidia-nccl-cu12==2.20.5 | nvidia-nccl-cu12==2.19.3 |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |
| wasabi==1.1.2 | wasabi==1.1.3 |
| botocore==1.34.116 | botocore==1.34.117 |
| boto3==1.34.116 | boto3==1.34.117 |
| ruff==0.4.6 | ruff==0.4.7 |
| fonttools==4.52.4 | fonttools==4.53.0 |
| wasabi==1.1.2 | wasabi==1.1.3 |
| botocore==1.34.116 | botocore==1.34.117 |
| boto3==1.34.116 | boto3==1.34.117 |
| ruff==0.4.6 | ruff==0.4.7 |
| fonttools==4.52.4 | fonttools==4.53.0 |
